src/audio/ipcgtw.c

```diff
@@ -0,0 +1,395 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright 2023 Intel Corporation. All rights reserved.
+
+#include <sof/audio/component_ext.h>
+#include <sof/trace/trace.h>
+#include <sof/lib/uuid.h>
+#include <sof/lib/memory.h>
+#include <sof/ut.h>
+#include <rtos/init.h>
+#include <ipc4/ipcgtw.h>
+#include <ipc4/copier.h>
+
+static const struct comp_driver comp_ipcgtw;
+
+LOG_MODULE_REGISTER(ipcgtw, CONFIG_SOF_LOG_LEVEL);
+
+/* a814a1ca-0b83-466c-9587-2f35ff8d12e8 */
+DECLARE_SOF_RT_UUID("ipcgw", ipcgtw_comp_uuid, 0xa814a1ca, 0x0b83, 0x466c,
+		    0x95, 0x87, 0x2f, 0x35, 0xff, 0x8d, 0x12, 0xe8);
+
+DECLARE_TR_CTX(ipcgtw_comp_tr, SOF_UUID(ipcgtw_comp_uuid), LOG_LEVEL_INFO);
+
+/* Host communicates with IPC gateways via global IPC messages. To address a particular
+ * IPC gateway, its node_id is sent in message payload. Hence we need to keep a list of existing
+ * IPC gateways and their node_ids to search for a gateway host wants to address.
+ */
+struct ipcgtw_data {
+	union ipc4_connector_node_id node_id;
+	struct comp_dev *dev;
+	struct list_item item;
+
+	/* IPC gateway buffer size comes in blob at creation time, we keep size here
+	 * to resize buffer later at ipcgtw_params().
+	 */
+	uint32_t buf_size;
+};
+
+/* List of existing IPC gateways */
+static struct list_item ipcgtw_list_head = { 0 };
+
+/* Incoming IPC gateway message */
+struct ipc4_ipc_gateway_cmd_data {
+	/* node_id of the target gateway */
+	union ipc4_connector_node_id node_id;
+	/* Payload (actual size is in the header extension.r.data_size) */
+	uint8_t payload[0];
+};
+
+/* Reply to IPC gateway message */
+struct ipc4_ipc_gateway_cmd_data_reply {
```

Comment:
can we use one convention?

ipc_gateway
ipc_gtw
ipcgtw

Pick one.

Reply:
I'm consistent here. I use ipcgtw everywhere except for IPC4 structs defined in spec. For those they are declared in C code as in spec but with ipc4_ prefix and with C convension. For example, scruct IpcGatewayCmd in spec is called ipc4_ipc_gateway_cmd in code. Hence, there are ipcgtw in some places (mostly in function names, variables) and ipc_gateway or something else in IPC4 structs and enums.